third_party/move/move-vm/types/src/loaded_data/instance_uid.rs

```diff
@@ -0,0 +1,126 @@
+// Copyright © Aptos Foundation
+// SPDX-License-Identifier: Apache-2.0
+
+use dashmap::{
+    mapref::entry::Entry::{Occupied, Vacant},
+    DashMap,
+};
+use derivative::Derivative;
+use std::{
+    cmp::{Eq, Ordering},
+    hash::{Hash, Hasher},
+    sync::{Arc, Weak},
+};
+
+// TODO: add tests.
+// TODO: break into files.
+// TODO: micro benchmarks.
+
+pub struct GlobalCache<'a, T: Eq + Hash> {
+    instance_to_compressed: DashMap<Arc<T>, Weak<InstanceUIDInner<'a, T>>>,
+}
+
+// Type wrapping Arc with overloaded methods to consider the underlying pointer.
+#[derive(Debug)]
+struct ArcKeyByPtr<T> {
+    key: Arc<T>,
+}
+
+impl<T: Eq + Hash> ArcKeyByPtr<T> {
+    fn new(key: Arc<T>) -> Self {
+        Self { key }
+    }
+}
+
+impl<T: Eq + Hash> PartialEq for ArcKeyByPtr<T> {
+    fn eq(&self, other: &Self) -> bool {
+        // TODO: https://github.com/rust-lang/rust/issues/106447, also
+        // TODO: should we allocate int ID and compare that?
+        Arc::ptr_eq(&self.key, &other.key)
+    }
+}
+impl<T: Eq + Hash> Eq for ArcKeyByPtr<T> {}
+impl<T: Eq + Hash> Ord for ArcKeyByPtr<T> {
+    fn cmp(&self, other: &Self) -> Ordering {
+        Arc::as_ptr(&self.key).cmp(&Arc::as_ptr(&other.key))
+    }
+}
+impl<T: Eq + Hash> PartialOrd for ArcKeyByPtr<T> {
+    fn partial_cmp(&self, other: &Self) -> Option<Ordering> {
+        Some(self.cmp(other))
```

Comment:
Why not use `as_ptr` here, too?  For BTree* that would be a big win.

Reply:
We need to collectively decide whether we want to break the orderings in BTree*.